iio/iio_app/iio_app.c

```diff
@@ -413,33 +373,80 @@ int32_t iio_app_run_with_trigs(struct iio_ctx_attr *ctx_attrs,
 	}
 
 	iio_init_param.devs = iio_init_devs;
-	iio_init_param.nb_devs = nb_devs;
-	iio_init_param.trigs = trigs;
-	iio_init_param.nb_trigs = nb_trigs;
-	iio_init_param.ctx_attrs = ctx_attrs;
-	iio_init_param.nb_ctx_attr = nb_ctx_attr;
-	status = iio_init(iio_desc, &iio_init_param);
+	iio_init_param.nb_devs = app_init_param.nb_devices;
+	iio_init_param.trigs = app_init_param.trigs;
+	iio_init_param.nb_trigs = app_init_param.nb_trigs;
+	iio_init_param.ctx_attrs = app_init_param.ctx_attrs;
+	iio_init_param.nb_ctx_attr = app_init_param.nb_ctx_attr;
+
+	status = iio_init(&application->iio_desc, &iio_init_param);
 	if(status < 0)
 		goto error;
 
 	free(iio_init_devs);
 
-	do {
-		status = iio_step(*iio_desc);
-	} while (true);
+	*app = application;
+
+	return 0;
 error:
-	status = print_uart_error_message(&uart_desc, uart_init_par, status);
+	free(application);
+
+	status = print_uart_error_message(&uart_desc,
+					  &app_init_param.user_uart_init_params, status);
 	return status;
 }
 
-int32_t iio_app_run(struct iio_ctx_attr *ctx_attrs, uint32_t nb_ctx_attr,
-		    struct iio_app_device *devices, uint32_t len)
+/**
+ * @brief Start an IIO application
+ *
+ * Configuration for communication is done through iio_app_init_param
+ * @param app - the iio application parameters
+ * @return 0 on success, negative value otherwise
+ */
+int iio_app_run(struct iio_app_desc *app)
 {
-	struct iio_desc	*iio_desc;
-	void *irq_desc = NULL;
+	int status;
 
-	return iio_app_run_with_trigs(ctx_attrs, nb_ctx_attr, devices, len, NULL, 0,
-				      irq_desc, &iio_desc);
+	do {
+		status = iio_step(app->iio_desc);
+		if (status)
+			return status;
+		if (app->post_step_callback) {
+			status = app->post_step_callback(app->arg);
+			if (status)
+				return status;
+		}
+	} while (true);
 }
 
+/**
+ * @brief Remove resources allocated by the IIO application
+ *
+ * Configuration for communication is done through iio_app_init_param
+ * @param app - the iio application parameters
+ * @return 0 on success, negative value otherwise
+ */
+int iio_app_remove(struct iio_app_desc *app)
+{
+	int ret;
+
+#if defined(ADUCM_PLATFORM) || (defined(XILINX_PLATFORM) && !defined(PLATFORM_MB)) || defined(STM32_PLATFORM)
+	ret = no_os_irq_ctrl_remove(app->irq_desc);
+	if (ret)
+		return ret;
 #endif
+
+	// ret = no_os_uart_remove(app->iio_desc->uart_desc);
```

Comment:
why is this commented out ?

Reply:
A separate member of iio_app_desc was needed.